src/endpoints/sources.js

```diff
@@ -105,10 +131,45 @@ const canonicalizeDataset = (pathBuilder) => (req, res, next) => {
  */
 const ifDatasetExists = async (req, res, next) => {
   if (!(await req.context.dataset.exists())) throw new NotFound();
-  next();
+  return next();
```

Comment:
The added `return` does make this a candidate for tail call optimization¹, but the reason I added the `return` here is because `ifDatasetExists()` went from being called directly by Express (which really only deals in sync functions, even with the `@awaitjs` wrapper) to being called by our `contentTypesProvided()` (which deals in async functions). The latter passes a `next()` function which is async (unlike Express'), so it needs awaiting:

https://github.com/nextstrain/nextstrain.org/blob/9b14ff8340b6b1de45a1e147bb44445f0f6f456a/src/negotiate.js#L12-L16

https://github.com/nextstrain/nextstrain.org/blob/9b14ff8340b6b1de45a1e147bb44445f0f6f456a/src/negotiate.js#L55-L69
 
That awaiting can happen explicitly in the handler with `await next()`, or implicitly by `return next()`. The implicit form is still compatible with Express' own `next()`, so by using it `ifDatasetExists` can still be used in either place.

¹ Though I don't know that any JS VM actually does tail call optimization.

Reply:
Ahh yes...  promises not returning async functions is such an easy bug to walk into and, in my experience, a tricky one to debug. The comment inside `contentTypesProvided` is spot on, but I think we need to go one step further and explicitly document each middleware function (in `./src/endpoints/sources.js`) because there we have two different types of middleware-like functions with subtle but important differences:

* `setNarrative` is correctly documented as "Express middleware" and so it can simply call `next()`, which it does.
* `ifNarrativeExists` docstring states "Express middleware function" but it's actually a handler we call ourselves (in the sense that express isn't calling it, we are calling it from an express-called middleware layer). This means it needs `return next()`, which it does.
* `setSource` (or the function returned from it) is only used as express middleware, yet uses `return next()`
* ditto for `setGroupSource`
* etc

Alternatively, why not simply enforce `return next()` in all middleware-like functions? (There exists middleware that wants to do work _after_ the chain is complete, but I don't think we have any.)

P.S. (from memory) express 5 is going to support async middleware out of the box, so that should make our lives easier.

Reply:
> Alternatively, why not simply enforce `return next()` in all middleware-like functions? (There exists middleware that wants to do work _after_ the chain is complete, but I don't think we have any. If we did, we could use `await next()` and make sure to add a comment explaining we're doing this on purpose.)

Yes, it is safe for middleware to always use `return next()` if you don't need to continue processing after calling `next()`. I didn't think it was worth changing everywhere to add `return` (instead opting to add it incidentally or as necessary), but I can do that if we want.

> P.S. (from memory) express 5 is going to support async middleware out of the box, so that should make our lives easier.

Yeah… but Express development over the past few years seems to be limited to minimal maintenance. While there are [alphas of Express 5](https://github.com/expressjs/express/releases?q=%225.0.0%22&expanded=true), it's not clear to me it'll ever get beyond that.